Consume control capacity

kdelee opened the pull request:
<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: "Consider capacity of node used for controlling a job before selecting and starting a job, and consume capacity on that instance. If there is no instance with control capacity, a job will remain in pending."
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Addresses https://github.com/ansible/awx/issues/10694
Replaces https://github.com/ansible/awx/pull/11651 
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 19.5.2.dev67+gbcba14e53e
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
This PR focuses exclusively on implementing https://github.com/ansible/awx/issues/10694

This implementation enforces the requirement for having all control and hybrid nodes be members of the `'controlplane'` instance group.

The tests did not previously comply with this requirement so I'm having to update a number of them. I have all but about 5 passing now. 


This PR:
  - selects a candidate controller node before we loop over instance groups
  - For non-container group jobs, if we end up selecting a hybrid node with enough capacity to do both the job and the control task, we prefer that node to do both
  - introduces setting `AWX_CONTROL_NODE_TASK_IMPACT` that is a constant integer (I use 5) amount of "task_impact" for when a node is the `controller_node` of a job
  - Selection of `task.execution_node` and `task.controller_node` and consumption of capacity in the in-memory capacity tracking happen before we go into `start_task`
- skip looping over `preferred_instance_groups` for project updates and system jobs. 

This PR does not:
- do any refactor of the in-memory capacity tracking mechanism.

rebeccahhh committed:
add test for control capacity consumption

add test to verify that when there are 2 jobs and only capacity for one
that one will move into waiting and the other stays in pending

rebeccahhh committed:
add test for hybrid node capacity consumption

assert that the hybrid node is used for both control and execution and
capacity is deducted correctly

kdelee committed:
Select control node before start task

Consume capacity on control nodes for controlling tasks and consider
remainging capacity on control nodes before selecting them.

This depends on the requirement that control and hybrid nodes should all
be in the instance group named 'controlplane'. Many tests do not satisfy that
requirement. I'll update the tests in another commit.

kdelee committed:
update tests to use controlplane

We don't start any tasks if we don't have a controlplane instance group

Due to updates to fixtures, update tests to set node type and capacity
explicitly so they get expected result.

shanemcd commented on awx/settings/development.py:
```diff
@@ -21,6 +21,8 @@
 from .defaults import *  # NOQA
 
+# How much capacity controlling a task costs a node
+AWX_CONTROL_NODE_TASK_IMPACT = 5
```
What's the difference between this variable and `AWX_CONTROL_PLANE_TASK_IMPACT` in `defaults.py`?

kdelee replied:
I agree this should not be necessary, I just did this because one test was failing on not picking up the setting in `defaults.py`. I can revert this part and try and get to the bottom of that

kdelee replied:
finally realized these were just two different words :D. fixed/deduplicated now

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -600,16 +626,17 @@ def reap_jobs_from_orphaned_instances(self):
-        self.graph[instance_group]['consumed_capacity'] += task.task_impact
+        self.graph[instance_group]['consumed_capacity'] += impact
         for capacity_type in ('control', 'execution'):
             if instance is None or instance.node_type in ('hybrid', capacity_type):
```
If you are consuming capacity for controller_node selection, then `capacity_type` should be "control", not whatever type the unified job is.

Otherwise your approach to call consume_capacity twice is valid and matches my expectations.

kdelee replied:
When we call this we pass the controller_node as the `instance`, so it will be hybrid or control. We don't reference the task.capacity_type here, rather we make decision based on the instance.node_type.

For hybrid node:
  - it will update:
    - `consumed_capacity`
    -  `consumed_execution_capacity`  
    -  `consumed_control_capacity`
    
for control node:
  - it will update:
    -  `consumed_capacity`
    -  `consumed_control_capacity`

Is that not right for hybrid nodes?
If that is not right for hybrid nodes, then are we also doing the wrong thing for jobs? Because anytime this is called w/ a hybrid node we'll deduct from both.

AlanCoding replied:
you're right. I was thinking that the `task` capacity type played into this, but it doesn't, because this is instance-oriented.

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -283,38 +283,12 @@ def start_task(self, task, rampart_group, dependent_tasks=None, instance=None):
                 schedule_task_manager()
+            # at this point we already have control/execution nodes selected for the following cases
             elif rampart_group.is_container_group:
                 task.instance_group = rampart_group
```
you might as well remove the `elif` because there's no difference between that and the `else`

kdelee replied:
done

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -510,13 +487,49 @@ def process_pending_tasks(self, pending_tasks):
+            remaining_control_capacity = self.get_remaining_capacity('controlplane', capacity_type='control')
+            if remaining_control_capacity < settings.AWX_CONTROL_PLANE_TASK_IMPACT:
+                logger.debug(f"Skipping task {task.log_format} in pending, not enough capacity left on controlplane to control new tasks")
+                continue
```
I'd vote to delete this condition. I was making notes myself to delete the similar condition for `execution_node` selection. If the IG doesn't have capacity, then the `fit_task_to_most_remaining_capacity_instance` / `find_largest_idle_instance` should fail to find a node. I know I asked you to back out your large refactor for now, but I want to hold onto the important critical thinking that came from it.

kdelee replied:
ok I've done this

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -510,13 +487,49 @@ def process_pending_tasks(self, pending_tasks):
+                logger.debug(f"Skipping task {task.log_format} in pending, not enough capacity left on controlplane to control new tasks")
+                continue
+
+            task.controller_node = control_instance.hostname
```
I have some concern about assigning here. I thought maybe a local variable for `controller_node` would be best, and then assign right before `start_task` is called. You could still do this, and it's not necessarily a bad idea, but it may force some surprising follow-on changes because I think it could get saved when _execution_ capacity falls short.

AlanCoding replied:
Next comment, if you get to this line and `task.capacity_type == 'control'`, then you should assign `execution_node` to the same value as `controller_node`, put it in the `controlplane` IG, start task, and then `continue`. Project updates and system jobs should not search over preferred instance groups.

kdelee replied:
did the second thing already. I will look at setting `task.controller_node` later

I _do_ think that is a bit paranoid though, because we dont call `task.save()` until the bottom of `start_task` ....which we never get to if we don't have execution capacity

AlanCoding replied:
What I was thinking of was at the end if this method we do `UnifiedJob.objects.bulk_update(tasks_to_update_job_explanation, ['job_explanation'])`, but now that I paste that I realize it won't update other fields, lol.

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -510,13 +487,49 @@ def process_pending_tasks(self, pending_tasks):
+            task.controller_node = control_instance.hostname
+
             for rampart_group in preferred_instance_groups:
                 if task.capacity_type == 'execution' and rampart_group.is_container_group:
```
if you accept my prior comment, `task.capacity_type == 'execution'` is always true, and can be deleted in cases like these.

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -510,13 +487,49 @@ def process_pending_tasks(self, pending_tasks):
+                    execution_instance = self.real_instances[control_instance.hostname]
+                    self.start_task(task, 'controlplane', task.get_jobs_fail_chain(), execution_instance)
+                    found_acceptable_queue = True
+                    break
```
Again, if you accept my prior comment about `capacity_type` being control, then the `if` condition here becomes unreachable and all of this block is deleted.

kdelee committed:
Fixes for accounting of control capacity consumed

Update method is used to account for currently consumed capacity for
instance groups in the in-memory capacity tracking data structure we initialize in
after_lock_init and then update via calculate_capacity_consumed (both in
task_manager.py)

Also update fit_task_to_instance to consider control impact on instances

Trust that these functions do the right thing looking for a
node with capacity, and cut out redundant check for the whole group's
capacity per Alan's reccomendation.

kdelee committed:
Refactor now redundant code

Deal with control type tasks before we loop over the preferred instance
groups, which cuts out the need for some redundant logic.

Also, fix a bug where I was missing assigning the execution node in one case!

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -532,9 +543,22 @@ def process_pending_tasks(self, pending_tasks):
                 ) or InstanceGroup.find_largest_idle_instance(self.graph[rampart_group.name]['instances'], capacity_type=task.capacity_type)
 
                 if execution_instance or rampart_group.is_container_group:
```
If it's a container group, then this code is unreachable, so delete `or rampart_group.is_container_group`

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -510,13 +485,49 @@ def process_pending_tasks(self, pending_tasks):
+                continue
+
+            control_instance = InstanceGroup.fit_task_to_most_remaining_capacity_instance(
+                task, self.graph['controlplane']['instances']
```
It looks like this needs something else. `fit_task_to_most_remaining_capacity_instance` will search for a suitable instance for `task.task_impact`, but what you need here is the controller_node impact instead (constant of 5 for most jobs).

kdelee commented on awx/main/scheduler/task_manager.py:
```diff
@@ -510,9 +486,39 @@ def process_pending_tasks(self, pending_tasks):
+                    # As in other places, we accept an idle instance if the node with most capacity (the control node we already selected)
+                    # does not have enough capacity.
+                    logger.debug(f"Not enough control capacity on {control_instance} to run {task.log_format}")
+                    break
```
@AlanCoding I think I should acutally call `continue` here and add
```
                task.log_lifecycle("needs_capacity")
                job_explanation = gettext_noop("This job is not ready to start because there is not enough available control capacity.")
```
what do you think of that

shanemcd replied:
I think you are correct, otherwise we're ejecting too early out of the `for task in pending_tasks` loop.

AlanCoding replied:
I don't think you need any of L501-L508. What question is this asking that wasn't already asked above when `control_instance` was set?

If the control plane doesn't have capacity for the control action (rewording your variable name), then `control_instance` will be `None` and you will have hit the prior `continue`. If all nodes had less than 5 capacity (however unlikely) but there were idle nodes, then the logic to define `control_instance` handles that too.

kdelee replied:
basically this catches the case when a node has 5 capacity but not 6 :laughing: maybe not worth it

AlanCoding replied:
Well, it wouldn't be worth it, but there's a better way to do that

```python
            control_instance = InstanceGroup.fit_task_to_most_remaining_capacity_instance(
                task, self.graph['controlplane']['instances'], impact=settings.AWX_CONTROL_NODE_TASK_IMPACT, capacity_type='control'
            ) or InstanceGroup.find_largest_idle_instance(self.graph['controlplane']['instances'], capacity_type='control')
```

You are specifically managing the impact here, `impact=settings.AWX_CONTROL_NODE_TASK_IMPACT`. If you have a "control" type task, just add 1 to that impact number you pass in.

kdelee replied:
we addressed this in the patch you helped me with

shanemcd commented on awx/main/managers.py:
```diff
@@ -243,7 +243,15 @@ def capacity_values(self, qs=None, tasks=None, breakdown=False, graph=None):
             impact = t.task_impact
-            if t.status == 'waiting' or not t.execution_node:
+            control_groups = []
+            if hasattr(t, 'controller_node') and t.controller_node:
```
The use of `hasattr` is a little suspicious here. Should we not just define a default property on `UnifiedJob`?

shanemcd replied:
It looks like it's a database field on UnifiedJob: https://github.com/ansible/awx/blob/ecf9a0827dc29338493ea7431393913029785458/awx/main/models/unified_jobs.py#L586

Why do we need this `hasattr`?

kdelee replied:
I ran into an exception in this area when working on this in my previous PR...but probably just me thrashing. I'll just call `t.controller_node`

kdelee replied:
done

kdelee committed:
set job explanation on tasks that need capacity

move the job explanation for jobs that need capacity to a function
so we can re-use it in the three places we need it.

kdelee commented on awx/main/managers.py:
```diff
@@ -243,7 +243,13 @@ def capacity_values(self, qs=None, tasks=None, breakdown=False, graph=None):
             impact = t.task_impact
-            if t.status == 'waiting' or not t.execution_node:
+            control_groups = []
+            if t.controller_node:
```
Ok @shanemcd I remember why I was doing the `hasattr(t, 'controller_node')`

Its because in the tests of this class, the Job objects don't have a `controller_node` attribute for some reason
```
FAILED awx/main/tests/unit/test_capacity.py::test_committed_capacity - AttributeError: 'Job' object has no attribute 'controller_node'
FAILED awx/main/tests/unit/test_capacity.py::test_running_capacity - AttributeError: 'Job' object has no attribute 'controller_node'
FAILED awx/main/tests/unit/test_capacity.py::test_offline_node_running - AttributeError: 'Job' object has no attribute 'controller_node'
FAILED awx/main/tests/unit/test_capacity.py::test_offline_node_waiting - AttributeError: 'Job' object has no attribute 'controller_node'
FAILED awx/main/tests/unit/test_capacity.py::test_RBAC_reduced_filter - AttributeError: 'Job' object has no attribute 'controller_node'
```

Looks like it is a mock class, so i'll update the tests

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -521,28 +529,35 @@ def process_pending_tasks(self, pending_tasks):
+                    # If our execution instance is a hybrid, prefer to do control tasks there as well.
+                    if (
+                        execution_instance.node_type == 'hybrid'
+                        and execution_instance.remaining_capacity >= task.task_impact + settings.AWX_CONTROL_PLANE_TASK_IMPACT
```
I think `and execution_instance.remaining_capacity >= task.task_impact + settings.AWX_CONTROL_PLANE_TASK_IMPACT` should be deleted.

Let me know if my comments haven't been making sense. My proposition is that control tasks (like project updates) and the extra penalty for hybrid nodes can fairly easily be handled in `fit_task_to_most_remaining_capacity_instance`, and I have example code for it in https://github.com/ansible/awx/compare/devel...AlanCoding:control_node_fast?expand=1#diff-6b5e2004df5844765d0ce3cb503debb4fb980d19fd436c4d9516b49a6685fd06

kdelee replied:
you addressed this in the patch I pushed

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -521,28 +529,35 @@ def process_pending_tasks(self, pending_tasks):
+                    control_instance.remaining_capacity = max(0, control_instance.remaining_capacity - settings.AWX_CONTROL_PLANE_TASK_IMPACT)
+                    task.log_lifecycle("controller_node_chosen")
+                    if control_instance != execution_instance:
+                        control_instance.jobs_running += 1
```
is `jobs_running` ever used for anything? I don't see why we should care about updating it for controller_node functions, but even if we did, this feels like the wrong thing to do. You could introduce a `jobs_controlling`, but just don't? Can we delete this?

EDIT: discussed, and this is the best option for satisfying the idle node check, given the tradeoffs

kdelee replied:
this is used when determining if an instance is idle, which is one of the possible ways it can be chosen

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -510,9 +486,41 @@ def process_pending_tasks(self, pending_tasks):
+                    # does not have enough capacity.
+                    logger.debug(f"Not enough control capacity on {control_instance} to run {task.log_format}")
+                    self.task_needs_capacity(task, tasks_to_update_job_explanation)
+                    continue
```
still want to get this `if` deleted, let me know if you want me to try and contribute a patch or anything.

AlanCoding committed:
project updates always run on the controlplane

Instance group ordering makes no sense on project updates because they
always need to run on the control plane.

Also, since hybrid nodes should always run the control processes for the
jobs running on them as execution nodes, account for this when looking for a
execution node.

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -485,6 +459,7 @@ def generate_dependencies(self, undeped_tasks):
     def process_pending_tasks(self, pending_tasks):
         running_workflow_templates = {wf.unified_job_template_id for wf in self.get_running_workflow_jobs()}
         tasks_to_update_job_explanation = []
+        controlplane_ig = InstanceGroup.objects.get(name='controlplane')
```
We're closing on finishing this up, so let me start with more nit-picky style stuff.

We should use `DEFAULT_CONTROL_PLANE_QUEUE_NAME` instead of the string "controlplane".

Also, I think you should hack this by adding onto `self` in `after_lock_init`. It does a loop over all instance groups, and you can just save this if the name matches this setting. Then you can reference `self.controlplane_ig` or whatever.

kdelee replied:
done

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -283,46 +283,20 @@ def start_task(self, task, rampart_group, dependent_tasks=None, instance=None):
-                    task.controller_node = task.execution_node
-                    task.log_lifecycle("controller_node_chosen")
-                logger.debug('Submitting job {} to queue {} controlled by {}.'.format(task.log_format, task.execution_node, task.controller_node))
+                queue_submitted_to = task.execution_node if task.execution_node else rampart_group.name
```
This smells funny. Instance group names aren't queues. All unified jobs (except for workflows and approval JTs, ugh) are submitted to the controller_node queue.

I think you should just rewrite this log

> Submitting {task.log_format} to control node {task.controller_node}

You could expand this to include the instance group name, and optionally the execution node, but execution node is blank for openshift, so pick your poison.

kdelee replied:
thanks, modified

kdelee committed:
fix misleading message

the variables and wording were both misleading, fix to be more accurate
description in the two different cases where this log may be emitted.

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -510,9 +488,36 @@ def process_pending_tasks(self, pending_tasks):
                 self.start_task(task, None, task.get_jobs_fail_chain(), None)
                 continue
 
+            # Determine if ther is control capacity for the task
```
ther -> there

kdelee committed:
use settings correctly

use settings.DEFAULT_CONTROL_PLANE_QUEUE_NAME instead of a hardcoded
name
cache the controlplane_ig object during the after lock init to avoid
an uneccesary query
eliminate mistakenly duplicated AWX_CONTROL_PLANE_TASK_IMPACT and use
only AWX_CONTROL_NODE_TASK_IMPACT

AlanCoding reviewed (approved):
As I approve, let me acknowledge that the managers.py code getting ugly. There are 2 spinoff issues from this, and that messy code will become a candidate for outright deletion as those are worked on.

kdelee committed:
add test for task.capacity_type = control

Test that control type tasks have the right capacity consumed and
get assigned to the right instance group

Also fix lint in the tests

kdelee committed:
jobs_running not accurate for control nodes

We can either NOT use "idle instances" for control nodes, or we need
to update the jobs_running property on the Instance model to count
jobs where the node is the controller_node.

I didn't do that because it may be an expensive query, and it would be
hard to make it match with jobs_running on the InstanceGroup which
filters on tasks assigned to the instance group.

This change chooses to stop considering "idle" control nodes an option,
since we can't acurrately identify them.

The way things are without any change, is we are continuing to over consume capacity on control nodes
because this method sees all control nodes as "idle" at the beginning
of the task manager run, and then only counts jobs started in that run
in the in-memory tracking. So jobs which last over a number of task
manager runs build up consuming capacity, which is accurately reported
via Instance.consumed_capacity

AlanCoding commented:
the test failure looks like it only needs a rebase

kdelee commented:
example of a job that waits on capacity for control then ends up running (Taken from a "hybrid-standalone" type deploy)
```
{"type": "job", "task_id": 1261, "state": "created", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:07.120251+00:00"}
{"type": "job", "task_id": 1261, "state": "acknowledged", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:07.974560+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:08.146437+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:08.529294+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:10.112022+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:17.020540+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:19.669164+00:00"}
{"type": "job", "task_id": 1261, "state": "controller_node_chosen", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "controller_node": "ec2-54-152-200-151.compute-1.amazonaws.com", "time": "2022-02-04T18:01:21.850781+00:00"}
{"type": "job", "task_id": 1261, "state": "execution_node_chosen", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "execution_node": "ec2-54-152-200-151.compute-1.amazonaws.com", "time": "2022-02-04T18:01:21.850976+00:00"}
{"type": "job", "task_id": 1261, "state": "waiting", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:21.883302+00:00"}
{"type": "job", "task_id": 1261, "state": "pre_run", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.250789+00:00"}
{"type": "job", "task_id": 1261, "state": "preparing_playbook", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.380333+00:00"}
{"type": "job", "task_id": 1261, "state": "running_playbook", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.470447+00:00"}
{"type": "job", "task_id": 1261, "state": "work_unit_id_received", "work_unit_id": "vD1UnMq9", "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.540987+00:00"}
{"type": "job", "task_id": 1261, "state": "work_unit_id_assigned", "work_unit_id": "vD1UnMq9", "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.591757+00:00"}
{"type": "job", "task_id": 1261, "state": "post_run", "work_unit_id": "vD1UnMq9", "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:25.701226+00:00"}
{"type": "job", "task_id": 1261, "state": "finalize_run", "work_unit_id": "vD1UnMq9", "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:25.781308+00:00"}
```

kdelee commented:
example of a job that waits on capacity for control then ends up running (Taken from a "hybrid-standalone" type deploy)
```
{"type": "job", "task_id": 1261, "state": "created", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:07.120251+00:00"}
{"type": "job", "task_id": 1261, "state": "acknowledged", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:07.974560+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:08.146437+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:08.529294+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:10.112022+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:17.020540+00:00"}
{"type": "job", "task_id": 1261, "state": "needs_capacity", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:19.669164+00:00"}
{"type": "job", "task_id": 1261, "state": "controller_node_chosen", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "controller_node": "my-hybrid-awx-instance.com", "time": "2022-02-04T18:01:21.850781+00:00"}
{"type": "job", "task_id": 1261, "state": "execution_node_chosen", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "execution_node": "my-hybrid-awx-instance.com", "time": "2022-02-04T18:01:21.850976+00:00"}
{"type": "job", "task_id": 1261, "state": "waiting", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:21.883302+00:00"}
{"type": "job", "task_id": 1261, "state": "pre_run", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.250789+00:00"}
{"type": "job", "task_id": 1261, "state": "preparing_playbook", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.380333+00:00"}
{"type": "job", "task_id": 1261, "state": "running_playbook", "work_unit_id": null, "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.470447+00:00"}
{"type": "job", "task_id": 1261, "state": "work_unit_id_received", "work_unit_id": "vD1UnMq9", "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.540987+00:00"}
{"type": "job", "task_id": 1261, "state": "work_unit_id_assigned", "work_unit_id": "vD1UnMq9", "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:22.591757+00:00"}
{"type": "job", "task_id": 1261, "state": "post_run", "work_unit_id": "vD1UnMq9", "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:25.701226+00:00"}
{"type": "job", "task_id": 1261, "state": "finalize_run", "work_unit_id": "vD1UnMq9", "template_name": "JobTemplate - SurveyParking\ubbea", "time": "2022-02-04T18:01:25.781308+00:00"}
```

we can see the hybrid node is quite busy both running and controlling jobs:
```
{
    "id": 1,
    "type": "instance",
    "url": "/api/v2/instances/1/",
    "related": {
        "named_url": "/api/v2/instances/my-hybrid-awx-instance.com/",
        "jobs": "/api/v2/instances/1/jobs/",
        "instance_groups": "/api/v2/instances/1/instance_groups/",
        "health_check": "/api/v2/instances/1/health_check/"
    },
    "uuid": "5b81710c-7518-42f4-b63d-7b13c0e90ab1",
    "hostname": "my-hybrid-awx-instance.com",
    "created": "2022-02-04T16:00:46.603459Z",
    "modified": "2022-02-04T16:00:46.603543Z",
    "last_seen": "2022-02-04T18:07:29.323864Z",
    "last_health_check": "2022-02-04T18:07:29.323864Z",
    "errors": "",
    "capacity_adjustment": "1.00",
    "version": "4.2.0",
    "capacity": 54,
    "consumed_capacity": 50,
    "percent_capacity_remaining": 7.41,
    "jobs_running": 6,
    "jobs_total": 356,
    "cpu": 2,
    "memory": 7835545600,
    "cpu_capacity": 8,
    "mem_capacity": 54,
    "enabled": true,
    "managed_by_policy": true,
    "node_type": "hybrid"
}
```

Also, from this same system running a number of jobs/project updates/workflows/inventory updates etc I have seen no tracebacks from any related changes in this PR (nothing from scheduler etc)

kdelee commented:
@AlanCoding I'm not seeing a fix for this in devel:
```
During handling of the above exception, another exception occurred:
/usr/lib64/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
awx/main/tests/functional/test_licenses.py:10: in <module>
    from pip.req import parse_requirements
E   ModuleNotFoundError: No module named 'pip.req'
```
That is what is causing my test failure

I'm not having luck re-creating that failure locally. As far as I can tell the import above this must be failing, and so its falling back to this old way to import it (pip < 10) 
https://github.com/pypa/pip/blob/ec8edbf5df977bb88e1c777dd44e26664d81e216/src/pip/_internal/commands/uninstall.py#L11
Looks like pip on main is doing the same way as us :confused:

kdelee commented:
figured out the bug...https://github.com/pytest-dev/pytest/pull/9609

I imagine we'll pin in devel and I'll rebase

AlanCoding commented:
Aside from rebasing, I think this is ready to merge.

kdelee committed:
Reduce default task impact for control nodes

This is something we can experiment with as far as what users
want at install time, but start with just 1 for now.

kdelee committed:
update capacity docs

Describe usage of the new setting and the concept of control impact.

kdelee commented:
Merging!

rebeccahhh reviewed (approved):
Approving even though I worked on this a bit, it looks great
